### PR TITLE
Implementation of per-brain reward thresholds.

### PR DESCRIPTION
**Pull request by drArthur**
Adding support for individual reward thresholds for each brain seperately.
In order to allow seperate thresholds for each brain, some new variables and accepted values are needed in the JSON file:
- added **"multibrain_reward"** for the **"measure"**
- thresholds have to be named exactly like their brain names.
- added **"threshold_amount"**, because it's a lot harder to infer the amount of thresholds from an arbitrary amount of arrays with arbitrary names.
- added **"threshold_operator"**, which can be set to either "or" or "and". It decides how all the thresholds are measured. In the case of "and", every brain's reward must be over their respective threshold and when it's "or", at least one brain's reward must be higher than the threshold.

 
Changes in trainer_controller.py:
- **_get_progress** returns a dictionary of Brain names coupled with their rewards, when the "measure" is set to "multibrain_reward".

Changes in curriculum.py:
- **__init__** additionally keeps track of **threshold_operator** and **threshold_amount**, changes the necessary values in the JSON file depending on whether the "measure" is set to "multibrain_reward" 

- **increment_lesson**, when "measure" is set to "multibrain_reward", iterates through each brain, reward pair and checks whether the lesson needs to be changed

An example for the JSON file:
```json
{
    "measure" : "multibrain_reward",
    "threshold_amount" : 2,
    "threshold_operator": "and",
    "SheepBrain": [0.5, 0.7],
    "WolfBrain": [0.2, 0.4],
    "min_lesson_length" : 4,
    "signal_smoothing" : true, 
    "parameters" : 
    {
        "CurrentLevel" : [1,2,3]
    }
}
```

### CODE REVIEW

**Comment by awjuliani** [comment]
Hi @drArthur,

Thanks for submitting this PR. Can you change the merge target to `develop` instead of `master`?

**Comment by drArthur** [comment]
Okay.
It's my first time doing anything like this, so sorry for my lack of experience :D

File: docs/Installation.md [hunk 1]
```diff
@@ -58,11 +58,7 @@ If you'd like to use Docker for ML-Agents, please follow
 
 ## Unity Packages
```
Comment (vincentpierre) [inline docs/Installation.md]: This commit should not be reverted.

File: docs/localized/zh-CN/docs/Installation.md [hunk 2]
```diff
@@ -56,11 +56,7 @@ Unity Assets。`python` 目录包含训练代码。
 
 ## Unity 包
```
Comment (vincentpierre) [inline docs/localized/zh-CN/docs/Installation.md]: This document should remain as it is

File: python/unityagents/curriculum.py [hunk 3]
```diff
@@ -29,14 +31,25 @@ def __init__(self, location, default_reset_parameters):
...
-                        'min_lesson_length', 'signal_smoothing']:
+            necessaryParams = ['parameters', 'measure',
+                        'min_lesson_length', 'signal_smoothing']
+            if(self.data['measure'] == "multibrain_reward"): 
+                self.smoothing_values = {}
+                necessaryParams.append('threshold_operator')
+                necessaryParams.append('threshold_amount')
+            else:    
+                necessaryParams.append('thresholds')
+            for key in necessaryParams:
```
Comment (vincentpierre) [inline python/unityagents/curriculum.py]: You call `self.data['measure'] ` before checking if the key is in the dictionary

**Comment by vincentpierre** [comment]
I am not a fan of having the brain names as keys of the JSON. We cannot guarantee that no one will call his brain "threshold_operator"...

File: python/unityagents/curriculum.py (hunk 3 above)
Comment (drArthur) [reply python/unityagents/curriculum.py]: Oh wow. 
Wasn't aware of this document being changed.
All the changes that I meant to make were in trainer_controller.py and curriculum only.

File: docs/localized/zh-CN/docs/Installation.md (hunk 2 above)
Comment (vincentpierre) [reply docs/localized/zh-CN/docs/Installation.md]: I think it is an error on our end, it is fixed now I think

**Comment by drArthur** [comment]
Made some updates to the code:

- implemented checking whether "measure" exists in JSON before checking if it equals to "multibrain_reward"
- now the threshold arrays for each brain are being searched inside "thresholds".
- the max_lesson_number is inferred from the threshold array with the least elements in it. Got rid of "threshold_amount" as a result of this.
Now the JSON file would look something like this:
```json
{
    "measure" : "multibrain_reward",
    "threshold_operator": "and",
    "thresholds":
    {
        "SheepBrain": [0.5, 0.7],
        "WolfBrain": [0.2, 0.4],
    },
    "min_lesson_length" : 4,
    "signal_smoothing" : true, 
    "parameters" : 
    {
        "CurrentLevel" : [1,2,3]
    }
}
```
I also tried to get rid of the problems created by making a pull request for develop while using master as the base for my edits, and merged develop into my branch. 
I'm still very new to working with github, so it is somewhat confusing at the moment...

**Comment by awjuliani** [comment]
Hi @drArthur,

Thanks for taking the time to work through this PR. In the next couple months we will be revamping the  Curriculum Learning system pretty significantly, and unfortunately the code here would have to be changed substantially in order to fit within the new system. That being said, we can definitely learn from it, and will be sure to acknowledge the contribution in the next ML-Agents release. Apologies that we couldn't merge it in.

**Pull request closed without merging**